pkg/innerring/innerring.go

```diff
@@ -352,15 +348,124 @@ func New(ctx context.Context, log *logger.Logger, cfg *viper.Viper, errChan chan
+		if !cfg.IsSet(committeeKey) {
+			return nil, fmt.Errorf("committee is not configured '%s'", committeeKey)
+		}
+		const storageKeyPrefix = morphConsensusKeyPrefix + "storage."
+		const useLevelDBKey = storageKeyPrefix + "use_leveldb"
+		if !cfg.IsSet(useLevelDBKey) {
+			return nil, fmt.Errorf("missing blockchain storage type flag '%s'", useLevelDBKey)
+		}
+		const storagePathKey = storageKeyPrefix + "path"
+		if !cfg.IsSet(storagePathKey) {
+			return nil, fmt.Errorf("missing path to the blockchain storage '%s'", storagePathKey)
+		}
+
+		useLevelDB, err := cast.ToBoolE(cfg.Get(useLevelDBKey))
+		if err != nil {
+			return nil, fmt.Errorf("invalid '%s' config (expected boolean): %w", useLevelDBKey, err)
+		}
+
+		committee, err := keys.NewPublicKeysFromStrings(cfg.GetStringSlice(committeeKey))
+		if err != nil {
+			return nil, fmt.Errorf("invalid committee configuration '%s': %w", committeeKey, err)
+		}
+
+		var cfgBlockchain blockchain.Config
+		cfgBlockchain.Wallet.Path = walletPath
+		cfgBlockchain.Wallet.Password = walletPass
+		cfgBlockchain.Storage.UseLevelDB = useLevelDB
+		cfgBlockchain.Storage.Path = cfg.GetString(storagePathKey)
+		cfgBlockchain.ErrorListener = errChan
+		cfgBlockchain.NetworkMagic = netmode.Magic(cfg.GetUint32(netMagicKey))
+		cfgBlockchain.Committee = committee
+
+		server.blockChain, err = blockchain.New(cfgBlockchain, blockchain.Options{
+			Logger:         log.Logger,
+			SeedNodes:      cfg.GetStringSlice(morphConsensusKeyPrefix + "seed_nodes"),
+			RPCEndpoints:   cfg.GetStringSlice(morphConsensusKeyPrefix + "rpc_endpoints"),
+			P2pEndpoints:   cfg.GetStringSlice(morphConsensusKeyPrefix + "p2p_endpoints"),
+			NotaryDisabled: cfg.GetBool(morphConsensusKeyPrefix + "notary_disabled"),
+		})
+		if err != nil {
+			return nil, fmt.Errorf("init internal blockchain: %w", err)
+		}
+
+		// don't move this code area within current function
+		{
+			log.Info("running blockchain...")
+			// we run node here instead of Server.Start because some Server components
+			// require blockchain to be run in the current function. Later it would be nice
+			// to separate the stages of construction and launch
+			err = server.blockChain.Run()
+			if err != nil {
+				return nil, fmt.Errorf("run internal blockchain: %w", err)
+			}
+
+			defer func() {
+				if err != nil {
+					server.blockChain.Stop()
+				}
+			}()
+		}
+
+		wsClient, err := server.blockChain.BuildWSClient(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("build WS client on internal blockchain: %w", err)
+		}
+
+		server.key = server.blockChain.LocalKey()
+		morphChain.key = server.key
+
+		server.morphClient, err = client.New(
+			server.key,
+			client.WithContext(ctx),
+			client.WithLogger(log),
+			client.WithSingleClient(wsClient),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("init internal morph client: %w", err)
+		}
+	} else {
+		// fallback to the pure RPC architecture
+		// TODO: think how we continue to support the case
```

Comment:
I think it's perfectly fine to keep it. The amount of code required to support it is minimal and yet it allows for easier migration and alternative setups in case they're really really really needed for some reason.